File: src/java/magmac/app/lang/node/Operation.java
package magmac.app.lang.node;

import magmac.app.compile.node.MapNode;
import magmac.app.compile.node.Node;
import magmac.app.compile.rule.LocatingRule;
import magmac.app.compile.rule.NodeRule;
import magmac.app.compile.rule.Rule;
import magmac.app.compile.rule.TypeRule;

record Operation(Value left, Operator operator, Value right) implements Value {

    public static Rule createOperationRule(Rule value, String type, String infix) {
        return new TypeRule(type, LocatingRule.First(new NodeRule("left", value), infix, new NodeRule("right", value)));
    }

    @Override
    public Node serialize() {
        return new MapNode();
    }
}

File: src/web/magmac/app/lang/node/Operation.ts
import { MapNode } from "../../../../magmac/app/compile/node/MapNode";
import { Node } from "../../../../magmac/app/compile/node/Node";
import { LocatingRule } from "../../../../magmac/app/compile/rule/LocatingRule";
import { NodeRule } from "../../../../magmac/app/compile/rule/NodeRule";
import { Rule } from "../../../../magmac/app/compile/rule/Rule";
import { TypeRule } from "../../../../magmac/app/compile/rule/TypeRule";
import { LazyRule } from "../../../../magmac/app/lang/LazyRule";
import { TypescriptValue } from "../../../../magmac/app/lang/web/TypescriptValue";
export class Operation {
	createOperationRule(operator : Operator, value : LazyRule) : Rule {return 0;;}
	serialize() : Node {return 0;;}
}
